pkg/auth/handler/sso/auth_handler.go

```diff
@@ -0,0 +1,295 @@
+func (f AuthHandlerFactory) NewHandler(request *http.Request) http.Handler {
+	h := &AuthHandler{}
+	inject.DefaultRequestInject(h, f.Dependency, request)
+	vars := mux.Vars(request)
+	h.ProviderName = vars["provider"]
+	h.Provider = h.ProviderFactory.NewProvider(h.ProviderName)
+	h.SSOSetting = h.ProviderFactory.Setting()
+	return h
+}
+
+func (f AuthHandlerFactory) ProvideAuthzPolicy() authz.Policy {
+	return policy.AllOf()
+}
+
+// AuthRequestPayload login handler request payload
+type AuthRequestPayload struct {
+	Code         string
+	Scope        sso.Scope
+	EncodedState string
+}
+
+// Validate request payload
+func (p AuthRequestPayload) Validate() error {
+	if p.Code == "" {
+		return skyerr.NewInvalidArgument("Authorization Code is required", []string{"code"})
+	}
+
+	if p.EncodedState == "" {
+		return skyerr.NewInvalidArgument("EncodedState is required", []string{"state"})
+	}
+
+	return nil
+}
+
+// AuthHandler decodes code response and fetch access token from provider.
+//
+// curl http://localhost:3000/sso/<provider>/auth_handler?code=<code>&state=<state>
+//
+// For ux_mode is 'ios' or 'android',
+// it creates a 302 response, and Location points to:
+// myapp://user.skygear.io/sso/{provider}/auth_handler?result=
+//
+// Fox ux_mode is 'web_redirect',
+// it creates a 302 response, and Location points to: sso_callback_url
+// and set cookie in the response.
+//
+// For ux_mode is 'web_popup',
+// it will render a html page and set cookie in the response.
+//
+type AuthHandler struct {
+	TxContext               db.TxContext                `dependency:"TxContext"`
+	AuthContext             coreAuth.ContextGetter      `dependency:"AuthContextGetter"`
+	OAuthAuthProvider       oauth.Provider              `dependency:"OAuthAuthProvider"`
+	PasswordAuthProvider    password.Provider           `dependency:"PasswordAuthProvider"`
+	AuthInfoStore           authinfo.Store              `dependency:"AuthInfoStore"`
+	RoleStore               role.Store                  `dependency:"RoleStore"`
+	TokenStore              authtoken.Store             `dependency:"TokenStore"`
+	AuthHandlerHTMLProvider sso.AuthHandlerHTMLProvider `dependency:"AuthHandlerHTMLProvider"`
+	ProviderFactory         *sso.ProviderFactory        `dependency:"SSOProviderFactory"`
+	Provider                sso.Provider
+	SSOSetting              sso.Setting
+	ProviderName            string
+}
+
+func (h AuthHandler) DecodeRequest(request *http.Request) (handler.RequestPayload, error) {
+	payload := AuthRequestPayload{}
+	q := request.URL.Query()
+	payload.Code = q.Get("code")
+	payload.Scope = strings.Split(q.Get("scope"), " ")
+	payload.EncodedState = q.Get("state")
+
+	return payload, nil
+}
+
+func (h AuthHandler) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	// reference from APIHandlerToHandler
+	var err error
+
+	defer func() {
+		if err != nil {
+			response := handler.APIResponse{
+				Err: skyerr.MakeError(err),
+			}
+			handler.WriteResponse(rw, response)
```

Comment:
Since this api is called by user browser which is not an api call, for all the errors we can show the error message directly like `validateCallbackUR` error? 

Would like to generalize the error handling of this handler, so we don't need special handling for error of `validateCallbackURL`. Thought?

Reply:
Maybe not? in `WriteResponse` it uses `nextSkyerr.ErrorDefaultStatusCode(response.Err)` to generate corresponding http error code from different `skyerr`, if we generalize error here, it could be `http.StatusBadRequest` for all errors.

Reply:
I think what we need to decide is the format of error response, json or plan text. If we want plan text, we can still use `nextSkyerr.ErrorDefaultStatusCode(response.Err)` to convert the status code. But if we use `WriteResponse` it also convert the response to json response.